ooni/nettests/blocking/web_connectivity.py

```diff
@@ -0,0 +1,512 @@
+            log.exception(exc)
+            log.err("Failed to lookup the resolver IP address")
+
+
+    def inputProcessor(self, filename):
+        """
+        This is a specialised inputProcessor that also supports taking as
+        input a csv file.
+        """
+        def csv_generator(fh):
+            for row in csv.reader(fh):
+                yield row[0]
+
+        def simple_file_generator(fh):
+            for line in fh:
+                l = line.strip()
+                # Skip empty lines
+                if not l:
+                    continue
+                # Skip comment lines
+                elif l.startswith('#'):
+                    continue
+                yield l
+
+        try:
+            fh = open(filename)
+            line = fh.readline()
+            # Detect the line of the citizenlab input file
+            if line.startswith("url,"):
+                generator = csv_generator(fh)
+            else:
+                fh.seek(0)
+                generator = simple_file_generator(fh)
+            for i in generator:
+                if not i.startswith("http"):
+                    i = "http://{}/".format(i)
+                yield i
+        finally:
+            fh.close()
+
+    def setUp(self):
+        """
+        Check for inputs.
+        """
+        if self.localOptions['url']:
+            self.input = self.localOptions['url']
+        if not self.input:
+            raise Exception("No input specified")
+
+        self.report['client_resolver'] = self.resolverIp
+        self.report['dns_consistency'] = None
+        self.report['body_length_match'] = None
+        self.report['headers_match'] = None
+        self.report['status_code_match'] = None
+
+        self.report['accessible'] = None
+        self.report['blocking'] = None
+
+        self.report['control_failure'] = None
+        self.report['http_experiment_failure'] = None
+        self.report['dns_experiment_failure'] = None
+
+        self.report['tcp_connect'] = []
+        self.report['control'] = {}
+
+        self.hostname = urlparse(self.input).netloc
+        if not self.hostname:
+            raise Exception("Invalid input")
+
+        self.control = {
+            'tcp_connect': {},
+            'dns': {
+                'ips': []
+            },
+            'http_request': {
+                'body_length': -1,
+                'failure': None,
+                'status_code': -1,
+                'headers': {}
+            }
+        }
+        if isinstance(self.localOptions['backend'], dict):
+            self.web_connectivity_client = WebConnectivityClient(
+                settings=self.localOptions['backend']
+            )
+        else:
+            self.web_connectivity_client = WebConnectivityClient(
+                self.localOptions['backend']
+            )
+
+    def experiment_dns_query(self):
+        log.msg("* doing DNS query for {}".format(self.hostname))
+        return self.performALookup(self.hostname)
```

Comment:
Currently this doesn't track CNAME records. I will add support for that as well.